x-pack/filebeat/input/awss3/sqs_acker.go

```diff
@@ -0,0 +1,212 @@
+	processingErr   error
+	Handles         []s3ObjectHandler
+	keepaliveCancel context.CancelFunc
+	keepaliveWg     *sync.WaitGroup
+	msgHandler      sqsProcessor
+	log             *logp.Logger
+}
+
+func NewEventACKTracker(ctx context.Context, deletionWg *sync.WaitGroup) *EventACKTracker {
+	ctx, cancel := context.WithCancel(ctx)
+	acker := &EventACKTracker{
+		ID:                ackerIDCounter.Inc(),
+		ctx:               ctx,
+		cancel:            cancel,
+		deletionWg:        deletionWg,
+		mutex:             new(sync.Mutex),
+		EventsAcked:       atomic.NewUint64(0),
+		EventsDropped:     atomic.NewUint64(0),
+		EventsPublished:   atomic.NewUint64(0),
+		EventsToBeTracked: atomic.NewUint64(0),
+	}
+
+	go func() {
+		t := time.NewTicker(500 * time.Microsecond)
+		defer t.Stop()
+
+		for {
+			<-t.C
+			if !acker.FullyTracked() {
+				continue
+			}
+
+			acker.cancelAndFlush()
+			return
+		}
+	}()
+
+	return acker
+}
+
+func (a *EventACKTracker) cancelAndFlush() {
+	a.cancel()
+	a.FlushForSQS()
+}
+
+// MarkSQSProcessedWithData Every call after the first one is a no-op
+func (a *EventACKTracker) MarkSQSProcessedWithData(msg *types.Message, publishedEvent uint64, receiveCount int, start time.Time, processingErr error, handles []s3ObjectHandler, keepaliveCancel context.CancelFunc, keepaliveWg *sync.WaitGroup, msgHandler sqsProcessor, log *logp.Logger) {
+	// We want to execute the logic of this call only once, when the ack mutex was locked on init
+	if a.EventsToBeTracked.Load() > 0 {
+		return
+	}
+
+	a.msg = msg
+	a.EventsToBeTracked = atomic.NewUint64(publishedEvent)
+	a.ReceiveCount = receiveCount
+	a.start = start
+	a.processingErr = processingErr
+	a.Handles = handles
+	a.keepaliveCancel = keepaliveCancel
+	a.keepaliveWg = keepaliveWg
+	a.msgHandler = msgHandler
+	a.log = log
+}
+
+func (a *EventACKTracker) FullyTracked() bool {
+	eventsToBeTracked := a.EventsToBeTracked.Load()
+	if eventsToBeTracked == 0 {
+		return false
+	}
+
+	// This is eating its own tail: we should check for dropped+published, but then we won't wait for acked.
+	// Acked might not be equal to published?
+	return a.EventsDropped.Load()+a.EventsAcked.Load() == eventsToBeTracked
+}
+
+// FlushForSQS delete related SQS message
+func (a *EventACKTracker) FlushForSQS() {
+	// Stop keepalive visibility routine before deleting.
+	a.keepaliveCancel()
+	a.keepaliveWg.Wait()
+
+	eventsPublished := a.EventsPublished.Load()
+	if eventsPublished > 0 && eventsPublished == a.EventsAcked.Load() {
+		err := a.msgHandler.DeleteSQS(a.msg, a.ReceiveCount, a.processingErr, a.Handles)
+		if err != nil {
+			a.log.Warnw("Failed deleting SQS message.",
+				"error", err,
+				"message_id", *a.msg.MessageId,
+				"elapsed_time_ns", time.Since(a.start))
+		} else {
+			a.log.Debugw("Success deleting SQS message.",
+				"message_id", *a.msg.MessageId,
+				"elapsed_time_ns", time.Since(a.start))
+		}
+	} else {
+		a.log.Infow("Skipping deleting SQS message, not all events acked.",
```

Comment:
Messages should still be deleted even if some events are dropped. For example a user might configure their processors to filter messages with a particular tag, but this doesn't mean that anything is wrong or that processing of that SQS message is not really "done." Until the client itself is closed, the Beats pipeline will only drop events that should be permanently dropped based on user settings.

Reply:
> Messages should still be deleted even if some events are dropped. For example a user might configure their processors to filter messages with a particular tag, but this doesn't mean that anything is wrong or that processing of that SQS message is not really "done." Until the client itself is closed, the Beats pipeline will only drop events that should be permanently dropped based on user settings.

Can you please clarify between "dropped", "published" and "acked"?

I assumed the following
- dropped: discarded by a processor in beats. not sent to the output
- published: sent to the output
- acked: output returned an positive response of receiving the event